### ISSUE TITLE
Inefficient use of DynamoDB during Journal Recovery

### ISSUE THREAD

**Issue by rmmeans** [akka/akka-persistence-dynamodb-1.x#106]
### Versions used 

akka-persistence-dynamodb version: `1.1.0`

### Expected Behavior

DynamoDB stores items that have a `sort key` on the same underlying partition as the `Partition Key` for performance reasons. Because this library makes good use of the `sort key` being a numeric type, Dynamo is able to optimize retrievals when you use ranging queries, etc.  

I would expect that this library, even though it is sharding out the given persistence ID into groups of 100 events per partition sequence, that when it asks for the journal replay on that partition sequence number, that it does so efficiently using Dynamo DB consistent read with a Query operation on the partition key and either not providing the sort key, or providing the range for what values to get within the sort key. This will result in the most efficient use of RCU (read capacity units) on Dynamo.

### Actual Behavior

During Journal recovery the replay finds all the range of sequence numbers that it must replay from the journal. In this scenario, we're going to assume there is no snapshotting happening - so if the persistent actor is not in memory, a full replay of the journal must occur. After the sequence numbers are discovered that need to be replayed (this operation is OK in terms of RCU cost to Dynamo), the library then proceeds to ask for each item using `BatchGetItem` with a consistent read. This is a huge problem in terms of RCU cost on the journal. Per the[ Dynamo Documentation](https://aws.amazon.com/dynamodb/pricing/provisioned/):
```
one RCU can perform one strongly consistent read request per second
```
This means that for a given persistent actors journal, we will need one RCU of capacity for each event on the journal! 

We have seen this in practice as we have some persistent actors with 8,000+ events. During a replay, we hit DynamoDB throttling for just a single persistent actor as Dynamo is not designed for this kind of inconsistent workloads. You can see this happening in the following screenshot:

During a large journal replay, we quickly exhaust the capacity of RCU available (this table is currently scaled to 8,000 provisioned RCU and has 4 underlying partitions according to AWS Premium support). The `BatchGetItem` requests start failing due to using all available RCU capacity, and those requests then get retried by the AWS SDK. The throttled requests are indicated by the red line.

![image](https://user-images.githubusercontent.com/3507512/144860321-79bd265a-7bbe-40b8-8403-5b3e2894a94a.png)

We've seen this issue cascade such that our persistent actors will fail to write anything to the journal because the journal was not able to replayed at all - so many throttles occurred that AWS SDK bubbled the exception up, which caused akka persistence to result in a journal error and we'll see the error logs from this location in: [EventSourced.scala Lines 144 - 173](https://github.com/akka/akka/blob/v2.5.32/akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala#L144-L173) show up with both a failure in the middle of a journal replay, but also failure even getting the first event!

### Incoming Fix

I have a PR coming in shortly that aims to address this issue. After the sequence numbers for replay are discovered, we use the Dynamo DB Query operation on each partition sequence number to get the events from that partition rather than asking for each item individually from Dynamo via `BatchGetItem`.  DynamoDB Query with sort keys was designed exactly for this use case. In practice I was able to reduce RCU capacity on a demo project down to about 1/10th the previous capacity needed. For instance, a persistent actor with 10,000 events took somewhere around ~1000 RCU to replay now instead of 10,000 RCU. 

Based on how Dynamo performs, best practice would be to always use a snapshotting plugin and avoid replaying journals as it has very bad performance characteristics. I would tend to recommend Dynamo snapshots need to happen every 100 events so-as to keep Dynamo utilization more flat.

**Comment by johanandren** [comment]
I guess this was fixed by #107 
@coreyoconnor should I do a 1.1.2 release with the improvement or does it need more testing first?

**Comment by coreyoconnor** [comment]
Afaict the current master is good to be released. I've done one, small, system test. Unfortunately, all I've had time for. :/

Anyways: yes please. If you can do a release that'd be great.

**Comment by coreyoconnor** [comment]
Closing due to fix in #107

**Comment by johanandren** [comment]
FYI: 1.1.2 including this fix on its way to maven central now
